src/trust_anchor_util.rs

```diff
@@ -16,25 +16,33 @@
 
 use std;
 use super::{Error, TrustAnchor};
-use super::cert::{EndEntityOrCA, parse_cert};
+use super::cert::{EndEntityOrCA, Cert, parse_cert};
+use super::cert::certificate_serial_number;
+use super::der;
 use untrusted;
 
+fn trust_anchor_from_cert<'a>(cert: Cert<'a>) -> TrustAnchor<'a> {
+    TrustAnchor {
+        subject: cert.subject.as_slice_less_safe(),
+        spki: cert.spki.as_slice_less_safe(),
+        name_constraints: cert.name_constraints
+                              .map(|nc| nc.as_slice_less_safe())
+    }
+}
+
 /// Interprets the given DER-encoded certificate as a `TrustAnchor`. The
 /// certificate is not validated. In particular, there is no check that the
 /// certificate is self-signed or even that the certificate has the cA basic
 /// constraint.
 pub fn cert_der_as_trust_anchor<'a>(cert_der: untrusted::Input<'a>)
                                     -> Result<TrustAnchor<'a>, Error> {
     // XXX: `EndEntityOrCA::EndEntity` is used instead of `EndEntityOrCA::CA`
-    // because we don't have a refernce to a child cert, which is needed for
+    // because we don't have a reference to a child cert, which is needed for
     // `EndEntityOrCA::CA`. For this purpose, it doesn't matter.
-    let cert = try!(parse_cert(cert_der, EndEntityOrCA::EndEntity));
-    Ok(TrustAnchor {
-        subject: cert.subject.as_slice_less_safe(),
-        spki: cert.spki.as_slice_less_safe(),
-        name_constraints: cert.name_constraints
-                              .map(|nc| nc.as_slice_less_safe())
-    })
+    parse_cert(cert_der, EndEntityOrCA::EndEntity)
+        .map(trust_anchor_from_cert)
+        .or_else(|err| parse_cert_v1(cert_der)
+                 .or(Err(err)))
```

Comment:
Given the discussion in #17, it seems like we should just use the new parser for all trust anchors. That would simplify the code in this PR. It won't completely solve #17, but fixing #17 would be just one step away.